controllers/release/adapter.go

```diff
@@ -495,29 +529,47 @@ func (a *adapter) createRoleBindingForClusterRole(clusterRole string, releasePla
 	return roleBinding, nil
 }
 
-// finalizeRelease will finalize the Release being processed, removing the associated resources.
-func (a *adapter) finalizeRelease() error {
-	pipelineRun, err := a.loader.GetReleasePipelineRun(a.ctx, a.client, a.release)
+// finalizeRelease will finalize the Release being processed, removing the associated resources. The pipelineRuns are optionally
+// deleted so that EnsureReleaseProcessingResourcesAreCleanedUp can call this and just remove the finalizers, but
+// EnsureFinalizersAreCalled will remove the finalizers and delete the pipelineRuns. If the pipelineRuns were deleted in
+// EnsureReleaseProcessingResourcesAreCleanedUp, they could be removed before all the tracking data is saved.
+func (a *adapter) finalizeRelease(delete bool) error {
+	// Cleanup Tenant Processing Resources
+	tenantPipelineRun, err := a.loader.GetReleasePipelineRun(a.ctx, a.client, a.release, metadata.TenantPipelineType)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	err = a.cleanupProcessingResources(tenantPipelineRun, nil)
 	if err != nil {
 		return err
 	}
 
-	if pipelineRun != nil {
-		// The finalizer could still exist at this point in the case of the PipelineRun not having succeeded at the time
-		// of finalizing the Release.
-		if controllerutil.ContainsFinalizer(pipelineRun, metadata.ReleaseFinalizer) {
-			patch := client.MergeFrom(pipelineRun.DeepCopy())
-			removedFinalizer := controllerutil.RemoveFinalizer(pipelineRun, metadata.ReleaseFinalizer)
-			if !removedFinalizer {
-				return fmt.Errorf("finalizer not removed")
-			}
-			err := a.client.Patch(a.ctx, pipelineRun, patch)
-			if err != nil {
-				return err
-			}
+	if delete && tenantPipelineRun != nil {
+		err = a.client.Delete(a.ctx, tenantPipelineRun)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
 		}
+	}
 
-		err = a.client.Delete(a.ctx, pipelineRun)
+	// Cleanup Managed Processing Resources
+	managedPipelineRun, err := a.loader.GetReleasePipelineRun(a.ctx, a.client, a.release, metadata.ManagedPipelineType)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	roleBinding, err := a.loader.GetRoleBindingFromReleaseStatus(a.ctx, a.client, a.release)
+	if err != nil && !errors.IsNotFound(err) && !strings.Contains(err.Error(), "valid reference to a RoleBinding") {
+		return err
+	}
+
+	err = a.cleanupProcessingResources(managedPipelineRun, roleBinding)
+	if err != nil {
+		return err
+	}
+
+	if delete && managedPipelineRun != nil {
+		err = a.client.Delete(a.ctx, managedPipelineRun)
```

Comment:
It would be cleaner if the blocks deleting the pipelines would be inside the `cleanUpProcessingResource`. You can even move there the piece getting the pipeline.

Reply:
I added it to the function comment why it is this way, but I will explain in the other comment about deleting them